packages/loader/container-loader/src/connectionStateHandler.ts

```diff
@@ -147,19 +147,21 @@ class ConnectionStateCatchup extends ConnectionStateHandlerPassThrough {
         switch (value) {
             case ConnectionState.Connected:
                 assert(this._connectionState === ConnectionState.CatchingUp, 0x3e1 /* connectivity transitions */);
-                assert(this.catchUpMonitor !== undefined,
-                    0x3e2 /* catchUpMonitor should always be set if pendingClientId is set */);
-                this.catchUpMonitor.on("caughtUp", this.transitionToConnectedState);
+                // Create catch-up monitor here (not earlier), as we might get more exact info by now about how far
+                // client is behind through join signal. This is only true if base layer uses signals (i.e. audience,
+                // not quorum, including for "rea" connections) to make decisions about moving to "connected" state.
+                // In addition to that, in its current form, doing this in ConnectionState.CatchingUp is dangerous as
+                // we might get callback right away, and it will screw up state transition (as code outside of switch
+                // statement will overwrite current state).
```

Comment:
This comment may be too detailed to leave in the code - but it's definitely helpful when reviewing this PR.

Based on the structure of the layers, it feels very natural that CatchUpMonitor would be instantiated here not in CatchingUp case.  This layer says "you think you're connected? Not so fast - I need to ensure we catch up".

Reply:
Can't tell for all engineers, but I'll forget things like that over time, and may cause a bug. So, I prefer code to be well documented RE intentions - I'd rather keep it here.